org.eclipse.draw2d/src/org/eclipse/draw2d/FigureCanvas.java

```diff
@@ -462,6 +463,26 @@ private void scrollChildren(int dx, int dy, Rectangle blit, Point dest) {
 		}
 	}
 
+	private void scrollRedraw(int dx, int dy, Rectangle clientArea, Rectangle blit) {
+		Rectangle expose = clientArea.getCopy();
+		if (dx != 0) {
+			expose.width = Math.abs(dx);
+			if (dx < 0) { // Moving left?
+				expose.x += blit.width; // Move expose area right
+			}
+			redraw(expose.x, expose.y, expose.width, expose.height, true);
+		}
+		if (dy != 0) {
+			expose.x = clientArea.x;
+			expose.width = clientArea.width;
+			expose.height = Math.abs(dy);
+			if (dy < 0) { // Moving up?
+				expose.y += blit.height; // Move expose area down
+			}
+			redraw(expose.x, expose.y, expose.width, expose.height, true);
+		}
```

Comment:
And this block is now the new code? Which, as far as I understand, redraws the -now visible- area?

Reply:
Yes the problem of the old code was that a way to small area was redrawn. This resulted under windows under extreme drawing artifacts.

Reply:
I see... I can't extensively test but from what I've heard, it works. At worst, we paint more than is needed, so I  don't mind giving it a go.

Reply:
In the case when both x and y are scrolled then we draw the overlapping square to much. But I think calculating that is more effort and makes the code much more unreadable.